### PR TITLE
Move to event based text input

### PR DESCRIPTION
**Pull request by Susko3**
Closes issues:
- https://github.com/ppy/osu/issues/3125
- https://github.com/ppy/osu/issues/3326
- https://github.com/ppy/osu/issues/6338
- https://github.com/ppy/osu-framework/issues/1965

Closes PRs:
- https://github.com/ppy/osu-framework/pull/1742
- https://github.com/ppy/osu-framework/pull/4709
  - all of the ideas/changes from there have now been implemented

---

Turns out to be quite simple, with minimal changes to textbox.
Mirrors existing key-blocking behavior, just that `consumingText` is replaced by `RecentTextInput`.
`RecentTextInput` is more robust, allowing us to remove the dodgy modifier key checking logic.

**This PR is made with the following in mind:**
- a text event will always come before a key press that caused it
  - that key press should be be blocked/consumed
  - this ordering is guaranteed because text events are sent from the main/input thread, while key event are sent from the update thread
- it's not neccassray that a key press happens after a text event
- if a key is held, it's not guaranteed that every repeating text event will have exactly one repeating key event (one or the other may be more frequent)
  - this is because the OS handles text repeat, while `InputManager` handles key repeat

Tested working as expected on Windows and Android.

## Issues

**• Dead keys are not handled properly**

SDL limitation, we don't get a text event for those.
Dead keys are commonly activated with <kbd>AltGr</kbd>, possibly causing problems with keyboard shortcuts getting activated when they shouldn't.

**• Arrows (and other text editing keys) don't work while `RecentTextInput` is `true`**

Importantly, backspace is not affected.

Possible real-world example:
Hold <kbd>Shift</kbd>, press a key (eg. <kbd>W</kbd>), then try to select text with <kbd>Shift</kbd>+<kbd>&larr;</kbd>. Nothing will happen.
Selecting works fine if you release all the keys and then use <kbd>Shift</kbd>+<kbd>&larr;</kbd>.

This could probably be fixed by more complex logic in `OnKeyUp()`, but is unnecessary.
Setting `RecentTextInput = false` when all keys are released seems like a reasonable and simple solution that would cover >99% use cases.

---

None of these issues are a deal-breaker, and they are also present on `master`.

### CODE REVIEW

**Comment by bdach** [comment]
> this ordering is guaranteed because text events are sent from the main/input thread, while key event are sent from the update thread

Can you elaborate on this? I am not feeling this too much. In general "guaranteed event ordering" raises immediate red flags because the point of events is that their order shouldn't matter.

Is it really impossible to have something like the following happen?

```
input thread						update thread

receives key event
schedules handling of key event to update thread
							receives and handles key event
receives text input event
```

The above may be impractically rare, sure, but I want to know how strong the guarantees are here for certain.

**Comment by Susko3** [comment]
Have now done a lot of testing and arrived at the following.

@bdach
Looking at an isolated key press+text input event combo, there simply is no way that will happen. We get the two SDL events in the same `pollSDLEvents()` call, first `SDL_KEYDOWN` and then `SDL_TEXTINPUT`.
From my testing, there is at least an order of magnitude difference between the two paths.

Using `Stopwatch.ElapsedTicks` for counting from the time that the keyboard event is enqueued in `InputHandler`, it takes 30~50 ticks to enqueue the text event at `TextBox`. Compare that to the difference between the last input dequeuing (last chance to enqueue the keyboard input) and keydown arrving at `TextBox`: >400 ticks. Even if the keyboard key is enqueued just as the queue is emptying, there will always be enough time for the text event to reach `TextBox`.

But this leads us to the following problem:
Key events being so much delayed that a previous keyup will happen in-between the next text input and keydown, thereby resetting `RecentTextInput`.

Example of a failure:
```cs
textinput: j
KeyDownEvent: J
textinput: n
keyup: J   // J released, and all keys released, as reported by framework, but N is down according to SDL.
KeyDownEvent: N   // not handled by textbox
keyup: N
```

But this is very rare, happening 1-2 times in 3 minutes of keyboard mashing, and not once on a release build of visual tests.

Even if a single keyboard key goes trough, it's not such a bit deal. It probably won't be handled by anything.
For `TextBox`, it's important that it handles potential global keyboard shortcuts, and those are usually made up of multiple keys so no problems there, as we block until all keys are released.

**Comment by bdach** [comment]
If I'm honest my doubts are not assuaged too much by the above explanation. While the empirical measurements check out *now*, they may not do so anymore at any given point in the future if either we or SDL upstream changes literally anything in the event ordering/timing/dispatch/whatever.

The main question for me is: do we absolutely *have* to interleave keypress and text input streams like this? It's just always going to be ugly and shaky ground however you make it happen, because it goes against the entire event-based design of... most windowing frameworks.

In other words: I would *probably* rather have `TextBox` force-block some or all key events everywhere else when focused than have this solution. My proposal would avoid the problem entirely by not having to attempt to correlate text input events with keypress events at all. And if the user wants to do something else despite the textbox being focused, too bad, they have to drop the focus by pressing Escape, or clicking away, or otherwise. Not sure what others think of that though.

**Comment by Susko3** [comment]
> I would *probably* rather have TextBox force-block some or all key events everywhere else when focused than have this solution. 

I think it would be beneficial if we look at some usages of `TextBox`, particularly in environments where keyboard shortcuts are used.

- osu! editor setup screen, where you might want to save with Ctrl+S when you eg. change a difficulty name
- chat, where <kbd>Alt</kbd>+<kbd>1~9</kbd> is used for switching tabs, which [may collide](https://github.com/ppy/osu/issues/3125#issuecomment-912376996 ) with option key combinations on macOS
  - should probably change the shortcut for that anyways
- `FocusedTextBox`, used frequently across osu!, force-blocking here is not going to work
  - Could use different blocking logic. But i think that might be confusing as then some textboxes block keyboard shortcuts, while others don't.

**Comment by Susko3** [comment]
If you want, we could strike a middle-ground. Keep the current logic as-is in this PR, which should block most key events as expected. But in case `RecentTextInput == false` also have the old logic, which would force-block some key events.

**Comment by peppy** [comment]
Finally got around to looking at this one.

Firstly, I think this is a good direction to take. But I also agree the flag is a bit haphazard (that said, so is text input across various platforms...)

I've done a refactor pass, half to make sure I understand what's going on, and half to tidy up usage around the flag to hopefully make it easier on the palette? @bdach interested to see if the changes make this change any better for you.

@Susko3 can you read through my commit and see if there's anything you disagree with?

I'm tentatively approving because this does solve multiple issues and the only way we're going to know if it's robust enough is to get it in users' hands.

**Review by peppy — DISMISSED** [review]

**Comment by peppy** [comment]
Hmm, I intended to put it on the explicit scheduler to enforce order of execution. Any reason for using the drawable one instead?

**Comment by Susko3** [comment]
I've went trough the problematic case described in that comment. And I can safely say that it doesn't matter what scheduler is used.
The exact problem is fixed with both schedulers.

`OLD` denotes behaviour without the schedule, `NEW` with.

```cs
// many frames ago, in an update frame, text and keydown event for 'O' processed as expected

--- begin update frame

UserInputManager.Update()  // no key events

// it's important for the bug that the event happens in between the two Update() calls.
SDL: keyup 'O' + keydown 'N' + text event 'N' (in that order)

TextBox.Update()   // text event processed:
                   // `textInputBlocking` set to true (no change, it was already true from the OnKeyDown)

--- end of update frame

// sleep until the next frame...

--- begin update frame

UserInputManager.Update() // the enqueued key events are processed:
 OnKeyUp: 'O'
  // OLD: set `textInputBlocking` to false
  // NEW: schedule `revertBlockingStateIfRequired()`
 OnKeyDown: 'N'
  // OLD: `textInputBlocking` is false, and no text events are scheduled, so this key event is not handled.
  // NEW: update the scheduler: run `revertBlockingStateIfRequired()`
  // this call to revertBlockingStateIfRequired() will not change the state of `textInputBlocking`, as a key is pressed
  // (I originally thought that this would set `textInputBlocking` to false, so that's why I used the drawable Scheduler.)

TextBox.Update()
 // if using drawable `Scheduler`, it's updated here. as you can see there is no difference
 // as revertBlockingStateIfRequired() will also not change the state here.

--- end of update frame
```

Now imagine what will happen if the second key is not <kbd>N</kbd> but something like <kbd>Enter</kbd>, or <kbd>F11</kbd> (hint: it'll be erroneously handled by textbox).

Anyways, the chances of this exact scenarion are very slim. Keys have to be pressed and released in the middle of an update frame. And worst case scenario (with the schedule) is that the key is handled when it shouldn't have been, so the user just presses it again.

And for the other cases, I feel like it should be beneficial to always have the text events have priority over `revertBlockingStateIfRequired()`, as we want to block eagerly -- the drawable `Scheduler` is better for that. We are already blocking more than necessary since we only clear the flag once all keys are released.

**Comment by peppy** [comment]
I feel a lot better having everything setting that flag in one scheduler, but let's see what others say.

File: osu.Framework.Tests/Visual/UserInterface/TestSceneTextBoxKeyEvents.cs [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+            {
+                InputManager.ReleaseKey(Key.F);
+                textInput.Text("S");
+                InputManager.PressKey(Key.S);
+            });
+            AddAssert("KeyDown event consumed", () => textBox.KeyDownQueue.Dequeue() && textBox.KeyDownQueue.Count == 0);
+            AddAssert("user text consumed event", () => textBox.UserConsumedTextQueue.Dequeue() == "S" && textBox.UserConsumedTextQueue.Count == 0);
+
+            AddStep("release key", () => InputManager.ReleaseKey(Key.S));
+            AddAssert("keys repeated", () => textBox.KeyDownRepeatQueue.Count != 0);
```
Comment (bdach) [inline osu.Framework.Tests/Visual/UserInterface/TestSceneTextBoxKeyEvents.cs:206]: not sure if just me but this step name misled me the first time i read through this - the implication here is that the repeat events were *fired for* (this line) and *consumed* (next line) by the text box, while "keys repeated" made me think that somehow the textbox wasn't consuming the events

File: osu.Framework.Tests/Visual/UserInterface/TestSceneTextBoxKeyEvents.cs [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+                                                        textBox.CommittedTextQueue.Count == 0 &&
+                                                        textBox.CaretMovedQueue.Count == 0 &&
+                                                        textBox.ImeCompositionQueue.Count == 0 &&
+                                                        textBox.ImeResultQueue.Count == 0 &&
+                                                        textBox.KeyDownQueue.Count == 0 &&
+                                                        textBox.PlatformActionQueue.Count == 0 &&
+                                                        textBox.KeyDownRepeatQueue.Count == 0);
+        }
+
+        private void assertKeyEventsNotConsumedForAction(PlatformAction action)
```
Comment (bdach) [inline osu.Framework.Tests/Visual/UserInterface/TestSceneTextBoxKeyEvents.cs:228]: this method also confuses me, i believe it's trying to check that the platform action was handled as a platform action rather than the set of its constituent keys, but that method name is not spelling it out too well

**Review by bdach — COMMENTED** [review]
i dunno. while this still feels brittle, because none of the underlying systems were built to support our usecase, and therefore i will not be surprised if it breaks N months down the line because the ordering doesn't happen to work the same way it currently does anymore, i also recognise that fixing this is substantial effort, likely on the SDL side too, so i'm willing to look away if this is appropriately marked and signposted from miles away.

the tests are a good start to that but i feel like some extended comments as to what precisely is being worked around with the text scheduler and all that stuff would also be warranted.

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 3]
```diff
@@ -233,16 +238,11 @@ public virtual bool OnPressed(KeyBindingPressEvent<PlatformAction> e)
                     return true;
 
                 case PlatformAction.Paste:
-                    string pending = null;
-
-                    //the text may get pasted into the hidden textbox, so we don't need any direct clipboard interaction here.
-                    if (inputBound)
-                        pending = textInput.GetPendingText();
+                    if (textInputBlocking)
+                        // the text has been pasted into the hidden textbox, so we don't need any direct clipboard interaction here.
```
[the remark this answers is not in the document]
Comment (Susko3) [reply osu.Framework/Graphics/UserInterface/TextBox.cs:242]: I wish we could remove this logic, but it seems iOS is using a hidden textbox:

https://github.com/ppy/osu-framework/blob/60d627e258ad199773fb79363e7509cf21a9cfcb/osu.Framework.iOS/IOSGameView.cs#L104

Somebody with an iOS device needs to test and confirm whether pasting text does in fact paste it into the hidden textbox.
Test with both a physical and virtual keyboard (and paste gestures, if those work in games).

[the remark this answers is not in the document]
Comment (Susko3) [reply osu.Framework/Graphics/UserInterface/TextBox.cs:250]: Looking more into this, if we're blocking key events from text input, and Ctrl+V is a key event that happens at the same time as a text event, it'll already get blocked in `OnKeyDown`. So it'll never reach `PlatformActionContainer`.

This can probably be removed.

**Review by smoogipoo — APPROVED** [review]
Just leaving my review here to say that I have no huge issues with this as it is. I do also share the same concerns as others with multiple schedulers being used (`textInputScheduler` scheduling to the `Drawable` scheduler).

I will leave others to merge.

File: osu.Framework/Graphics/UserInterface/TextBox.cs [hunk 4]
```diff
@@ -133,6 +133,11 @@ public bool ReadOnly
         /// </summary>
         public event OnCommitHandler OnCommit;
 
+        /// <summary>
+        /// Scheduler used for scheduling text input events coming from <see cref="textInput"/>.
+        /// </summary>
+        private readonly Scheduler textInputScheduler = new Scheduler(() => ThreadSafety.IsUpdateThread, null);
```
[the remark this answers is not in the document]
Comment (peppy) [reply osu.Framework/Graphics/UserInterface/TextBox.cs]: Given its purpose, it might read better as a list of delegates, or anything that isn't a `Scheduler`. But using a scheduler does encapsulate all required queueing logic so I'm not sure.

**Comment by peppy** [comment]
Let's go with what we have here and see how things play out.
